kpi/tasks.py

```diff
@@ -140,3 +148,137 @@ def remove_old_versions():
+                s.submission_counts_all_time = data.get('submission_counts_all_time', 0)
+                s.current_period_submissions = data.get('current_period_submissions', 0)
+                s.billing_period_start = data.get('billing_period_start')
+                s.billing_period_end = data.get('billing_period_end')
+                s.snapshot_created_at = now
+                s.last_snapshot_run_id = run_id
+                to_update.append(s)
+            else:
+                s = BillingAndUsageSnapshot(
+                    organization_id=oid,
+                    effective_user_id=data.get('effective_user_id'),
+                    storage_bytes_total=data.get('storage_bytes_total', 0),
+                    submission_counts_all_time=data.get('submission_counts_all_time', 0),
+                    current_period_submissions=data.get('current_period_submissions', 0),
+                    billing_period_start=data.get('billing_period_start'),
+                    billing_period_end=data.get('billing_period_end'),
+                    snapshot_created_at=now,
+                    last_snapshot_run_id=run_id,
+                )
+                to_create.append(s)
+
+        # update/create in a transaction per chunk
+        with transaction.atomic():
+            if to_update:
+                for i in range(0, len(to_update), BULK_UPDATE_BATCH):
+                    batch = to_update[i: i + BULK_UPDATE_BATCH]
+                    BillingAndUsageSnapshot.objects.bulk_update(
+                        batch,
+                        fields=[
+                            'effective_user_id',
+                            'storage_bytes_total',
+                            'submission_counts_all_time',
+                            'current_period_submissions',
+                            'billing_period_start',
+                            'billing_period_end',
+                            'snapshot_created_at',
+                            'last_snapshot_run_id',
+                        ],
+                        batch_size=BULK_UPDATE_BATCH,
+                    )
+
+            if to_create:
+                for i in range(0, len(to_create), BULK_CREATE_BATCH):
+                    batch = to_create[i: i + BULK_CREATE_BATCH]
+                    BillingAndUsageSnapshot.objects.bulk_create(
+                        batch, batch_size=BULK_CREATE_BATCH
+                    )
+
+    # Delete stale rows (rows not touched by this run) in small batches
+    while True:
+        stale_ids = list(
+            BillingAndUsageSnapshot.objects
+            .filter(~Q(last_snapshot_run_id=run_id))
+            .values_list('pk', flat=True)[:DELETE_BATCH]
+        )
+        if not stale_ids:
+            break
+        BillingAndUsageSnapshot.objects.filter(pk__in=stale_ids).delete()
+
+    # Refresh materialized view
+    with connection.cursor() as cursor:
+        cursor.execute('REFRESH MATERIALIZED VIEW CONCURRENTLY user_reports_mv;')
```

Comment:
it seems that it does not work with `CONCURRENTLY` 

> ERROR:  cannot refresh materialized view "public.user_reports_mv" concurrently
> HINT:  Create a unique index with no WHERE clause on one or more columns of the materialized view.

but running without concurrenty will lock the view for ~ a 1 minute on production.